### ISSUE TITLE
Enhancement: Use Cap'n Proto instead of Messagepack

### ISSUE THREAD

**Issue by dumblob** [neovim/neovim#5526]
For the last 2 years as I follow the development of NeoVim, every single time I open the main NeoVim GitHub page and see *MessagePack*, I ask myself why didn't I already proposed [Cap'n Proto](https://capnproto.org/ ) as a better alternative.

Long story short, [Cap'n Proto](https://capnproto.org/ ) is basically the fastest possible binary serialization format. It's also quite easy to implement in [any language](https://capnproto.org/otherlang.html ).

I think it's worth switching to it as it will make NeoVim frontends faster and with lower latency (imagine interpreted languages like JavaScript working with MessagePack) and allow for higher efficiency in CPU utilization for larger buffer changes.

**Comment by mondico** [comment]
> imagine interpreted languages like JavaScript working with MessagePack

Are there any benchmarks showing that Cap'n Proto provides any benefit for an interpreted language like Javascript?

**Comment by justinmk** [comment]
Decoupling the serialization format might be worth a look some day. But at the moment serialization is not the bottleneck. There are substantially bigger gains to be had by being smarter about what changes are sent.

**Comment by bfredl** [comment]
https://capnproto.org/rpc.html is kinda interesting, but nothing really stops us from doing the same things on top of msgpack-rpc like #4568 (when references are implemented we should allow referencing previous results) or (from the boring-and-ugly-but-already-works-well department) by rplugins injecting a bit vimL (in the future lua) into nvim to do these things.

**Comment by justinmk** [comment]
Agreed. Pipelining seems more relevant to systems that (ahem) don't offer "remote eval" :)

In a plugin/extension/coordination ecosystem, stability and commonality is 10x more valuable than a moving target that improves things by 10% or even 200% (1000% is a different story). Let's see measurements first, and real-world experiences where msgpack is limiting some use case.

**Comment by dumblob** [comment]
I don't have any recent benchmarks for any JavaScript implementation of Cap'n Proto. I can just point out, that MessagePack is in this particular case (JavaScript implementation) slower than JSON (because of built-in JSON support in JavaScript interpreters), whereas JSON is in general known to be very slow serialization format (just a tiny demonstration: [BSON versus MessagePack](https://jsperf.com/json-bson-msgpack/5 )). So I'm sure the [Cap'n Proto JavaScript implementation](https://github.com/capnp-js/serialization ) is faster than MessagePack.

Cap'n Proto for Lua supports FFI and is very fast (as expected) - see https://blog.cloudflare.com/introducing-lua-capnproto-better-serialization-in-lua/ and the repository https://github.com/cloudflare/lua-capnproto .

The Cap'n Proto main architect also recommends the C implementation https://github.com/opensourcerouting/c-capnproto .

Of course if currently MessagePack is not a performance bottleneck, we shouldn't hurry with making serialization more efficient as the gain against MessagePack would be "only" about 6x for low level implementations ([C/C++](https://github.com/thekvs/cpp-serializers )) and 20x for implementation in interpreted languages (such as Lua with FFI).

But I agree, that it would make sense to decouple RPC from serialization, which Cap'n Proto allows (as well as MessagePack to be fair).

**Comment by purpleP** [comment]
@dumblob I don't know much about neovim internals and I'm not a maintainer, but I can point out a one thing that you seem to not taking into account.

MessagePack is schemaless and capnproto, protobuf etc are schema-based. And schemaless gives you one important property - better programmers productivity.
1. You don't need to write schema. That's obviously better.
2. You don't need to use generated code. That's not that obvious for some people. If you're using generated code for serialization for the most projects you **should** have a layer of abstraction above it. It tends to be fairly stupid level that for the most part copies the interface of generated code, but if you wouldn't do it you would be stuck with you serialization protocol for basically forever (if you can't or not willing to change all clients). I've seen projects that used protobuf generated code directly without a layer of abstraction. I bet they still using it although there are better alternatives now.
And this two parts gives programmer more productivity. And that's exactly the things that makes schemaless serialization less performant.

So you shoudn't think how much capnproto is faster than MessagePack but how much the speed is relevant in this case. So I would say that probably not only "we shouldn't hurry making serialization more efficient" but probably (not for me to decide) never do it at all, because of the cost of productivity.

**Comment by dumblob** [comment]
@purpleP good points, thanks for bringing them up.

>1. You don't need to write schema. That's obviously better.

Not in long term. It's better solely for the first attempt to "make it work somehow". In other words, schema-less significantly lowers the initial burden, but schema-based promotes stable, long-existing schema. With schema I explicitly say *"This is the way of thinking about the data and related performance in this project, accommodate yourself."* and with schema-less *"These are the data we wanted to publish today. Tomorrow we might want to publish different data, so don't be surprised by that and don't expect it to be performance-optimized."*.

>2. You don't need to use generated code. That's not that obvious for some people. If you're using generated code for serialization for the most projects you **should** have a layer of abstraction above it...

Layers of abstraction are just a matter of personal opinion. Your example with Protocol Buffers clearly shows how it's done in [Go](https://golang.org/ ) (as Protocol Buffers were developed to be performant in Go and to have neat minimal interface in Go). Namely "there is only one way of writing code and using the language" (including style, structuring, etc. - see e.g. the famous ["bug" about `}else{`](https://github.com/golang/go/issues/5440 ) and many more other rants about preprocessing, freedom for bindings and embedding, etc.). Milions of programmers love Go just for this prescribed structure - because of the nature of code being read 100x more than written (i.e. *conformity and resulting clarity over creativity*). So, I must say no, there is no objective need for writing an abstraction layer.

Anyway this discussion is irelevant, because it's very general (not directly related to NeoVim) and leads us astray. The question is though what main NeoVim developers think about the added value of schema or schema-less in long term.

**Comment by purpleP** [comment]
@dumblob I'm sorry, but I can't agree.
1. Yes, schema-based promotes stable schema etc. But so does discipline. I mean it's entirely up to team to change or not change the schema. You don't want users of api suffer from changes to schema - don't change it.

> These are the data we wanted to publish today. Tomorrow we might want to publish different data, so don't be surprised by that and don't expect it to be performance-optimized.

That's not true. At least not all truth. With schema-less it's "These are the data we want to publish today. If we want api to be stable we will tell you about it, if we're not - we wouldn't". So again, It's up to developer to specify whether api would change or not in the future. And the fact that they published their current schema have nothing to do with it. They still can change the schema entirely if they would like.

2. And layers of abstraction is not a matter of opinions. Look at this python code for protobufs

```python
class Person(message.Message):
  __metaclass__ = reflection.GeneratedProtocolMessageType

  class PhoneNumber(message.Message):
    __metaclass__ = reflection.GeneratedProtocolMessageType
    DESCRIPTOR = _PERSON_PHONENUMBER
  DESCRIPTOR = _PERSON

class AddressBook(message.Message):
  __metaclass__ = reflection.GeneratedProtocolMessageType
  DESCRIPTOR = _ADDRESSBOOK

import addressbook_pb2
person = addressbook_pb2.Person()
person.id = 1234
person.name = "John Doe"
person.email = "jdoe@example.com"
phone = person.phone.add()
phone.number = "555-4321"
phone.type = addressbook_pb2.Person.HOME
```
I can assure that if you wouldn't make a layer above this message layer if you would have to rewrite it, because in now way that's a pythonic code. So to limit effect of a change you **need** another layer.

**Comment by bfredl** [comment]
> With schema I explicitly say "This is the way of thinking about the data and related performance in this project, accommodate yourself." and with schema-less "These are the data we wanted to publish today. Tomorrow we might want to publish different data, so don't be surprised by that and don't expect it to be performance-optimized.".

With 0.1.6 and later we will formalize strong guarantees about rpc api stability. An API method's metadata must not change in any way, not even widening arguments or narrowing return values. Rather a new method must be added and the old one must be deprecated for least for one api level and only removed at a major release point. On top of this, schema based rpc might be an worthwhile _optimization_ but it will not provide any further stability or long-existingness guarantees, which already will be strong enough.

And the api must be an abstraction layer, because it is not only accessible through rpc, also through vimL and in the likely future though direct C calls with in-process embedding.

**Comment by mondico** [comment]
Nvim [provides a schema in machine readable form](https://neovim.io/doc/user/msgpack_rpc.html#rpc-api).

**Comment by dumblob** [comment]
@bfredl thank you for revealing the plans. Apparently the abstractions are in case of Neovim independent from the underlying serialization technology, which also renders the discussion about Cap'n Proto (not) needing writing an abstraction layer pointless.

@mondico wow, that's great. Nice to see that already now there is a need for machine readable schema. In case Cap'n Proto will be used at some point, the switch won't be so painful.

**Comment by lifranc** [comment]
Bringing in the info which was lacking in a duplicate I created to suggest using CBOR format for the rpc functionality in neovim (#10246).

Here's a comment which goes over what I recalled were the problems with msgpack, from someone explaining why IPLD used CBOR over msgpack:

 > Msgpack has some bonkerinos stuff about being unable to tell the difference between strings and bytes, iirc. This may not be true anymore -- I think it was eventually patched into a saner state (though not necessarily with an unambiguous migration path) -- but the ecosystemic uncertainty damage was done. (Discussions: msgpack/msgpack#121 , msgpack/msgpack#128 )
(So, yeah, "not canonical representation".)

Notable projects using CBOR include [matrix](https://fosdem.org/2019/schedule/event/matrix/) (as eventual JSON replacement), IPLD (IPFS backend), and [CoAP](https://coap.technology/)

Note that my suggestion mostly stemmed from my unproven belief that msgpack being an ancestor of CBOR would mean easy porting. The fork is over 5 years old now so this could be somewhat unfounded, but probably not wildly so.

**Comment by bfredl** [comment]
Those linked issues are 6-7 years old. All maintained msgpack libs implement the "new" spec with STR and EXT types.

**Comment by lifranc** [comment]
True, and I must take back the statement that msgpack refused to break, since they did in with v5 of the spec.
Hence it is understandably not enticing for nvim to migrate.

So I'll simply leave it at: I would develop on CBOR since there are other projects I wish to talk to that use it. That is all.

I also forgot to mention that dhall has a CBOR representation, which is what got me into this really. Dhall is probably the most innovative language I know of today. Maybe that counts for something.

**Comment by lifranc** [comment]
@bfredl Having non-seekable data is ... I guess the word isnt allowed here.
Imagine a world of interoperable, navigatable data ...
https://www.youtube.com/watch?v=PBo1lEZ_Iik

**Comment by lifranc** [comment]
I think this whole topic deserves a new issue, but I'll ask this time.

This is about the *next standard*. So the right question is:

#### What is the next UTF8 ?

Of course to perhaps most programmers this doesnt matter, but my work is about creating the most efficient computer systems possible. So (directly) supporting multiple data format is completly out of the question.

Message pack can't be the next data type since it doesn't have union, or sum types, which is a basic data type in modern computing systems. Im basing my first implementation on [bincode](https://github.com/servo/bincode) but of course it is only rust right now. CBOR just seems to be where everyone will settle on if only because IPLD and dhall are based on it. Dhall is the only language I use which has real code representation, and it uses a CBOR format, which supports sum type using a tuple structure I think, or something similar. Who knows the original author might have a word to say about this; he might use neovim too!

**Comment by justinmk** [comment]
We can't spend time chasing marginally improved formats, there are higher priorities right now. 

Msgpack is quite reasonable in the meantime, and it is a _reversible_ decision. There is zero reason for us to spin our wheels optimizing this until there is an obvious, unambiguously better alternative.

**Comment by lifranc** [comment]
> until there is an obvious, unambiguously better alternative.

That's exacly what I'm working on. So if a truck doesnt hit me it'll be bincode-based. So this issue now depends on a c++ port of [bincode](https://github.com/servo/bincode).

**Comment by maisiliym** [comment]
> We can't spend time chasing marginally improved formats, there are higher priorities right now.

True, the only sensible move is to a format which *cannot be improved further*. Capnproto pretty much fits the bill; improvements can be made of course but more on implementation details than the format itself; a format like capnproto has reached the limits/design-restraints of the hardware architecture of the day (8/16/32/64 bits words).

In other words, it looks like capnproto is actually worth it.
Otherwise, msgpack was a lot nicer choice than json/toml/yaml,other-utf8-nonsense.

This might be relevant to Li's sajban project, which at this point looks like it will use nvim as its first UI.
And lately he has been thinking about a capnproto-formated sajban (as opposed to bincode, which is rust-only for now. They are both rather similar in deserialized form).
This issue might reach the people who already have similar interests...

**Comment by bfredl** [comment]
> True, the only sensible move is to a format which cannot be improved further. Capnproto pretty much fits the bill; improvements can be made of course but more on implementation details than the format itself; a format like capnproto has reached the limits/design-restraints of the hardware architecture of the day (8/16/32/64 bits words).

For capnproto to be seriously considreded what would be needed is
(1) a clear overview fo what limitations of msgpack/msgpack-rpc it improves on
(2) some sketch how our API conventions (while _inspired_ by msgpack is not directly tied to it; in-process code can use same API without encode/decode) map to capnproto (or similar)

To be clear: this something which existing core contributors most likely would not be driving for either the 0.5 or 0.6 cycle, but we are happy to guide some one willing to de heavy lifting. but again, a clear analysis of the benefits (and risks) should be presented and discussed before coding.

**Comment by maisiliym** [comment]
Another reason why it isn't worth all this work; nvim uses lots of strings, which would need to be changed to enum(from rust-dialect, C uses different term) types to get most of the benefit. Which is also why LI gave up on capnproto; not enough benefits for all those legacy applications which did not have de/serialize optimization in mind when designing the data structure (the main reason to do this was to communicate with all popular frameworks). By the time a good data format comes, nvim will be about as useful as `ed`. So Li is weighting for closing this and focusing on multi-selection and multi-frontends, without which nvim is still a bit of a laughing stock.

**Comment by dumblob** [comment]
>Another reason why it isn't worth all this work; nvim uses lots of strings, which would need to be changed to enum(from rust-dialect, C uses different term) types to get most of the benefit. Which is also why LI gave up on capnproto; not enough benefits for all those legacy applications which did not have de/serialize optimization in mind when designing the data structure (the main reason to do this was to communicate with all popular frameworks).

I can't comment on the general effort needed, but I'd like to point out, that Rust has absolutely no issues with Cap'n Proto and than Rust's enums **are** fully supported by Cap'n Proto. So I'm not sure the issues are truly related to Nvim's "strings" or other data types.

**Comment by maisiliym** [comment]
> Rust's enums **are** fully supported by Cap'n Proto. 

That's unrelated to the last comment, which started with: 
> > Another reason why it isn't worth all this work

And then went on to explain the said 'why' ...

In other words; it's possible but a major waste of time.

**Comment by maisiliym** [comment]
Related: msgpack-rust in a bit of disrepair; https://github.com/3Hren/msgpack-rust/issues/255
It seems msgpack did not (yet) draw activation-energy-level interest rust-side, for a number of reasons, which include both CBOR and bincode.

Apparently, the CBOR debate was more smoke than fire in the end, but the person/people behind CBOR had better PR/marketing skills/friends in high places, and so the ecosystem was fragmented a few years ago, right around the time when neovim would have picked msgpack. Now, people seem to point/switch to CBOR/others, which is creating a msgpack exodus.

Here is in interesting excerpt, from [this blog post](https://diziet.dreamwidth.org/6568.html) (emphasis added)
"MessagePack seems to have been prompted into fixing the missing string type problem, but only by the threat of a fork. However, this fork went ahead even after MessagePack clearly accepted the need for a string type. MessagePack had a fixed protocol spec before the IETF did.

The continued pursuit of the IETF fork was ostensibly been motivated by a disapproval of the development process and in particular a sense that the IETF process was superior. [...]

*One lesson for the rest of us is that although being the upstream and nominally in charge of a project seems to give us a lot of power, it's wise to listen carefully to one's users and downstreams. Once people are annoyed enough to fork, the fork will have a life of its own.*"

And so it seems the lack of love for msgpack is unwarranted. CBOR offers little over it.

Note: the commenter is simply sharing and not engaging into an argumentation.
